lib/src/fixed_datetime_formatter.dart

```diff
@@ -0,0 +1,336 @@
+
+/// A formatter and parser for [DateTime] in a fixed format [String] pattern.
+///
+/// For example, calling
+/// `FixedDateTimeCodec('YYYYMMDDhhmmss').decodeToLocal('19960425050322')` has
+/// the same result as calling `DateTime(1996, 4, 25, 5, 3, 22)`.
+///
+/// The allowed characters are
+/// * `Y`	for “calendar year”
+/// * `M`	for “calendar month”
+/// * `D`	for “calendar day”
+/// * `E`	for “decade”
+/// * `C`	for “century”
+/// * `h`	for “clock hour”
+/// * `m`	for “clock minute”
+/// * `s`	for “clock second”
+/// * `S`	for “fractional clock second”
+///
+/// Note: Negative years are not supported.
+///
+/// Non-allowed characters in the format [pattern] are included when decoding a
+/// string, in this case `YYYY kiwi MM` is the same format string as
+/// `YYYY------MM`. When encoding a datetime, the non-format characters are in
+/// the output verbatim.
+///
+/// Note: this class differs from [DateFormat] in that here, the characters are
+/// treated literally, i.e., the format string `YYY` matching `996` would result
+/// in the same as calling `DateTime(996)`.
+class FixedDateTimeFormatter {
+  static const _powersOfTen = [1, 10, 100, 1000, 10000, 100000];
+  static const _validFormatCharacters = [
+    _yearCode,
+    _monthCode,
+    _dayCode,
+    _decadeCode,
+    _centuryCode,
+    _hourCode,
+    _minuteCode,
+    _secondCode,
+    _fractionSecondCode,
+  ];
+  static const _yearCode = 0x59; /*Y*/
+  static const _monthCode = 0x4D; /*M*/
+  static const _dayCode = 0x44; /*D*/
+  static const _decadeCode = 0x45; /*E*/
+  static const _centuryCode = 0x43; /*C*/
+  static const _hourCode = 0x68; /*H*/
+  static const _minuteCode = 0x6D; /*m*/
+  static const _secondCode = 0x73; /*s*/
+  static const _fractionSecondCode = 0x53; /*S*/
+
+  /// The format pattern string of this formatter, stored publicly in case the
+  /// user wants to retrieve it.
+  final String pattern;
+
+  /// Whether to create UTC [DateTime] objects when parsing.
+  ///
+  /// If not, the created [DateTime] objects are in the local time zone.
+  final bool isUtc;
+
+  final _blocks = _ParsedFormatBlocks();
+
+  /// Creates a new [FixedDateTimeFormatter] with the provided [pattern].
+  ///
+  /// The [pattern] interprets the characters mentioned in
+  /// [FixedDateTimeFormatter] to represent fields of a `DateTime` value. Other
+  /// characters are not special. If [isUtc] is set to false, the DateTime is
+  /// constructed with respect to the local timezone.
+  ///
+  /// There must at most be one sequence of each special character to ensure a
+  /// single source of truth when constructing the [DateTime], so a pattern of
+  /// `"CCCC-MM-DD, CC"` is invalid, because it has two separate `C` sequences.
+  FixedDateTimeFormatter(this.pattern, {this.isUtc = true}) {
+    int? currentCharacter;
+    var start = 0;
+    var characters = pattern.codeUnits;
+    for (var i = 0; i < characters.length; i++) {
+      var formatCharacter = characters[i];
```

Comment:
I'd just use `pattern.codeUnitAt(i);` here, and not introduce a `characters` wrapper.

Reply:
I didn't benchmark this (and don't know how the implementation looks like), and it doesn't really matter anyhow in the constructor, but I assumed `codeUnitAt` to be slower in a loop than extracting the `codeUnits` once.